### PR TITLE
Ws week picker preact

### PR DESCRIPTION
**Pull request by faboweb**
Week Picker:
image

PLZ COMPARE TO:
image

ATTENTION: I used a darker color for the headings then it is in the current date-picker (similar design). I propose this, because the contrast in the date-picker is too low for my taste. I already proposed this before when I build the week picker web component.

@wholesale-design-system/ux plz approve

### CODE REVIEW

**Comment by faboweb** [comment]
The date picker actually has a better contrast to the background using a more from the top shadow and borders.

**Comment by faboweb** [comment]
#20 is treating the arrow shadow issue

**Comment by myueksel** [comment]
Like this version with the gray background and the subtile border. Makes the needle (Pinöppel) more visible than before

File: src/ws-week-picker/ws-week-picker-calendar.js [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+ */
+const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
+/**
+ * Quick array of all month shown. There are two from the last year and two from the next year.
+ */
+const allMonths = [months[10], months[11]].concat(months).concat([months[0], months[1]]);
+
+/**
+ * Renders a week picker calendar component.
+ */
```
Comment (fokusferit) [inline src/ws-week-picker/ws-week-picker-calendar.js]: I would appreciate if you could add a short description about the props maybe? Like I did for WSHeader ß

File: src/ws-week-picker/ws-week-picker-calendar.js [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    };
+
+    const today = new Date(Date.now());
+    this.todayYear = today.getFullYear();
+    this.todayWeek = getWeekOfYear(today);
+  }
+
+  /**
+   * Show the previous year.
+   * @returns {void}
```
Comment (fokusferit) [inline src/ws-week-picker/ws-week-picker-calendar.js]: You don't have to do this :)

File: src/ws-week-picker/ws-week-picker.scss [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+    caption {
+      padding: 0.5rem 0;
+      color: $heron;
+      font-weight: 300;
+      font-size: $base-font-size;
+
+      > span {
+        cursor: pointer;
+        line-height: 2rem;
```
Comment (tonysaad) [inline src/ws-week-picker/ws-week-picker.scss]: I'd make it 1.75rem; and just remove the height so the arrow align with the year

File: src/ws-week-picker/ws-week-picker.scss [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        height: 2rem;
+      }
+      i {
+        position: relative;
+        top: 0.45rem;
```
Comment (tonysaad) [inline src/ws-week-picker/ws-week-picker.scss]: in addition to previous comment, Adding vertical-align would fix it
`&:before {
    vertical-align: top;
}`

File: src/ws-week-picker/ws-week-picker-calendar.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+   * @returns {JSX.Elements[]}
+  */
+  buildWeekRows(weeksPerMonth) {
+    // there are up to 5 weeks per month
+    return [0, 1, 2, 3, 4].map(weekIndex =>
+      <tr key={weekIndex}>
+        {
+          allMonths.map((month, monthIndex) => {
+            const weekInMonth = weeksPerMonth[monthIndex][weekIndex];
+            if (weekInMonth == null) return <td key={`${monthIndex}_${weekIndex}`}></td>;
```
Comment (fokusferit) [inline src/ws-week-picker/ws-week-picker-calendar.js]: As weekPerMonth is a Input what happens if it is "undefined" ?
Especially as we only have a check for weekInMonth being Null or not. Does it make sense to also capture 'undefined' here ?

File: src/ws-week-picker/ws-week-picker.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
+import {React, Component} from '../imports';
+import {WSWeekPickerCalendar} from './ws-week-picker-calendar';
+import './ws-week-picker.scss';
+
+/**
+ * Renders a week picker component.
+ */
```
Comment (fokusferit) [inline src/ws-week-picker/ws-week-picker.js]: again at least the onChange Handler / props should be documented somehow.

File: src/index.js [hunk 7]
```diff
@@ -9,6 +9,7 @@
  * - See {@link ws-date-picker/ws-date-picker.md|WSDatePicker}
  * - See {@link ws-dropdown/ws-dropdown.md|WSDropdown}
  * - See {@link ws-notification/ws-notification|WSNotification}
+ * - See {@link ws-week-picker/ws-week-picker|WSWeekPicker}
```
Comment (fokusferit) [inline src/index.js]: I'm sorry, could you please change this to {@link ws-week-picker/ws-week-picker.md|WSWeekPicker} with the ".md" Extension

**Comment by schluchter** [comment]
@faboweb I agree that the arrow-tip is more visible in this design. The subtle border is a great idea! What I don't quite like is that the background color of the pop-over seems "dirty white" relative to the paper in the background. 

Compromise: Slightly deeper shadow that bleeds out of to the top as well with a barely visible border around the box. Looks like this in Sketch:
rectangle 5

That way, we can keep the contrast of content to background higher.

**Comment by faboweb** [comment]
@tonysaad thx for implementing the fix :)

**Comment by faboweb** [comment]
@schluchter I am fine with either. Are you aligned with @myueksel ? Then I would implement it right away.

**Comment by myueksel** [comment]
Nice!!!

**Comment by faboweb** [comment]
Latest version:
image

Plz approve: @schluchter @myueksel @tonysaad @fokusferit

File: src/ws-week-picker/ws-week-picker-calendar.js [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+            const {week, year} = weekInMonth;
+            return (
+              <td className={(monthIndex < 2 || monthIndex > 13 ? 'off ' : '')
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker-calendar.js]: We had naming conventions defined were we said that state classes gets a prefix with `is-`  like `is-active` and modifying classes are prefixed with `mod-` like `mod-large`

File: src/ws-week-picker/ws-week-picker-calendar.js [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+      </tr>);
+  }
+
+  /**
+   * Renders the calendar.
+   * @returns {void}
+   */
+  render() {
+    const weeksPerMonth = [];
+    for (let i = -2; i <= 13; i++) {
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker-calendar.js]: Why do you start here from -2? A comment with an explanation why u choosed the numbers would be cool. Like getting +- 2 month more for better overview.

File: src/ws-week-picker/ws-week-picker-calendar.js [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+      weeksPerMonth.push(getWeeks(i, this.state.showingYear));
+    }
+    return (
+      <div className="ws-date-picker-calendar">
+        <table>
+          <caption>
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker-calendar.js]: Caption tag is not supported anymore in html 5

File: src/ws-week-picker/ws-week-picker-calendar.js [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+ * @param {number} week Week
+ * @param {number} year Year of the week
+ * @returns {{week: number, year: number}[]}
+*/
+function getWeeks(month, year) {
+  while (month > 11) {
+    ++year;
+    month = month - 12;
+  }
+  while (month < 0) {
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker-calendar.js]: couldn't you do something like `year -= Math.ceil(month / 12)`

File: src/ws-week-picker/ws-week-picker.js [hunk 12]
```diff
@@ -0,0 +1,128 @@
+import {React, Component} from '../imports';
+import {WSWeekPickerCalendar} from './ws-week-picker-calendar';
+import './ws-week-picker.scss';
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker.js]: please DON'T import the scss in the javascript. Import it from the scss files like demo/index.scss.

File: src/ws-week-picker/ws-week-picker.js [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        <input
+          value={this.state.selectedWeek != null ? `Week ${this.state.selectedWeek}, ${this.state.selectedYear}` : ''}
+          placeholder={'Please choose a week'}
+          onClick={() => this.toggleCalendar()}
+          readOnly
+        />
+        <i className={`icon icon-${this.state.show ? 'cross' : 'calendar'} icon16`}
+          onClick={() => this.toggleCalendar()}
+        />
+        {this.state.show ? <WSWeekPickerCalendar onChange={selection => this.onChange(selection)} selectedYear={this.state.selectedYear} selectedWeek={this.state.selectedWeek} /> : null}
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker.js]: Why do you wrap the icon but not this long line?

File: src/ws-week-picker/ws-week-picker.scss [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+      width: 0;
+      border-width: 0.5rem;
+      margin: 0 -8px;
+      border-bottom-color: $earl-gray;
+      bottom: 100%;
+      left: 1rem;
+    }
+    &:after {
+      border-width: calc(0.5rem - 1px);
+      margin: 0 -7px;
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker.scss]: Sometimes you use pixel and sometimes rem.

File: src/ws-week-picker/ws-week-picker.scss [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+      border-bottom-color: $athens-gray;
+    }
+  }
+
+  table {
+    table-layout: fixed;
+    display: inline-block;
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker.scss]: Idk really but does it make sense to set a fixed table layout and make the "table" then to an inline block? Does this work?

File: src/ws-week-picker/ws-week-picker.scss [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+    border: none;
+    font-size: $font-size-small;
+
+    caption {
+      padding: 0.5rem 0;
+      color: $heron;
+      font-weight: 300;
+      font-size: $base-font-size;
+
+      > span {
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker.scss]: span has the class name .prev, .next this is more descriptive as span.

File: src/ws-week-picker/ws-week-picker.scss [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+        cursor: pointer;
+        line-height: 1.75rem;
+      }
+      i {
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker.scss]: I think this should be .icon

File: src/ws-week-picker/ws-week-picker.scss [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+      th {
+        padding: 0 !important;
+        line-height: 2rem;
+      }
+    }
+    tbody {
+      @include table-padding(0, $padding, 0, $padding);
+
+      &:before, &:after {
+        content: "_";
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker.scss]: For what is this?

File: src/ws-week-picker/ws-week-picker.scss [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+      td {
+        padding: rem(2px) rem(2px) !important;
+      }
+    }
+    td, th {
+      text-align: center;
+      background-color: transparent !important;
+      border: none !important;
+
+      &.off, &.off .week {
```
Comment (fragsalat) [inline src/ws-week-picker/ws-week-picker.scss]: Should be .is-off

**Comment by myueksel** [comment]
Updated of CW picker... aligned to new spacings and styles (same as the date picker).

Invision:
https://zalando.invisionapp.com/d/main/#/console/10454474/235119440/preview

cw picker v1

**Comment by fokusferit** [comment]
@myueksel could you just put your last comment input into the following issue: #30 this here is a closed PR and it will be lost 😄

**Pull request merged**